src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java

```diff
@@ -208,4 +252,111 @@ else if (!testMessages.isEmpty()) {
         return feedback;
     }
 
+    /**
+     * Generates test cases from the given result's feedbacks & notifies the subscribing users about the test cases if they have changed. Has the side effect of sending a message
+     * through the websocket!
+     *
+     * @param buildResult from which to extract the test cases.
+     * @param exercise    the programming exercise for which the test cases should be extracted from the new result
+     */
+    public void extractTestCasesFromResultAndBroadcastUpdates(AbstractBuildResultNotificationDTO buildResult, ProgrammingExercise exercise) {
+        boolean haveTestCasesChanged = generateTestCasesFromBuildResult(buildResult, exercise);
+        if (haveTestCasesChanged) {
+            // Notify the client about the updated testCases
+            Set<ProgrammingExerciseTestCase> testCases = testCaseRepository.findByExerciseId(exercise.getId());
+            websocketMessagingService.sendMessage("/topic/programming-exercises/" + exercise.getId() + "/test-cases", testCases);
+        }
+    }
+
+    /**
+     * From a list of build run feedback, extract all test cases. If an already stored test case is not found anymore in the build result, it will not be deleted, but set inactive.
+     * This way old test cases are not lost, some interfaces in the client might need this information to e.g. show warnings.
```

Comment:
```suggestion
     * From a list of build run feedback, extract all test cases. If an already stored test case is not found anymore in the build result, it will not be deleted but set inactive.
     * This way old test cases are not lost, some interfaces in the client might need this information, e.g., to show warnings.
```